### PR TITLE
Bug 1847923 - Part 2: Add network middleware for review quality check

### PR DESCRIPTION
**Pull request by rahulsainani**
More details in the bug. 

### How
– Add `ReviewQualityCheckNetworkMiddleware` that handles network related actions. It delegates to `ReviewQualityCheckService` to fetch the product analysis and then maps to UI State, then dispatches an update action on the store.
– Remove fake data in `ReviewQualityCheckState` as the data is now coming from the network.

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1847923

### CODE REVIEW

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt:45]: This is deliberately on the main thread so it works right now, hence the draft. When [this](https://bugzilla.mozilla.org/show_bug.cgi?id=1848008) patch lands, we can update this to the default dispatcher.

Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt:45]: My understanding about this was incorrect above ☝️ It still throws `IllegalThreadStateException` when called on any other dispatcher.

Does [GeckoResult.then](https://searchfox.org/mozilla-central/source/mobile/android/geckoview/src/main/java/org/mozilla/geckoview/GeckoResult.java#567) expect to be called on the main thread so `IllegalThreadStateException` is not called? @calumozilla @owlishDeveloper 

Seems we have to `requestAnalysis` on `Dispatchers.Main` as done in [GeckoSitePermissionsStorage](https://searchfox.org/mozilla-mobile/source/firefox-android/android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/permission/GeckoSitePermissionsStorage.kt#134) and [GeckoCookieBannersStorage](https://searchfox.org/mozilla-mobile/source/firefox-android/android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/cookiebanners/GeckoCookieBannersStorage.kt#100)

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/di/ReviewQualityCheckMiddlewareProvider.kt
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/di/ReviewQualityCheckMiddlewareProvider.kt:44]: out of scope but could we inline ` reviewQualityCheckPreferences = ReviewQualityCheckPreferencesImpl(settings),` ?

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ProductAnalysisMapper.kt
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ProductAnalysisMapper.kt:63]: Plugging it again here 🙂 
```suggestion
private fun Highlight.highlightForType(highlightType: HighlightType) =
```
or
```suggestion
private fun Highlight.ForType(highlightType: HighlightType) =
```

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ProductAnalysisMapper.kt
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ProductAnalysisMapper.kt:77]: 👌

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt:45]: Yep, if we run this branch and change to Dispatchers.Default, the stack trace points to `GeckoResult.then`

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt:44]: nit: think we can lose this doc

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckAction.kt
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckAction.kt:30]: does it make sense to group this above with the other `ReviewQualityCheckAction` interfaces?

File: fenix/app/src/test/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckServiceImplTest.kt
Comment (t-p-white) [inline fenix/app/src/test/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckServiceImplTest.kt:33]: nit: could import this and remove the qualifiers

File: fenix/app/src/test/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckServiceImplTest.kt
Comment (t-p-white) [inline fenix/app/src/test/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckServiceImplTest.kt:28]: given the test WHEN fetch is called THEN fetch is called for the selected tab I wonder how useful this is

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ProductAnalysisMapper.kt
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ProductAnalysisMapper.kt:61]: Kotlin magic *chef's kiss*

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt:48]: nit: if it's only going to be these two that fetch the analysis, then it can be cleaned up slightly (the arrow on the newline is off putting to me and I can't explain why 😅 )
```suggestion
            ReviewQualityCheckAction.FetchProductAnalysis,
            ReviewQualityCheckAction.RetryProductAnalysis -> {
```

**Review by MozillaNoah — COMMENTED** [review]
quick round of nits/suggestions/comments as I bounce around. I don't have any main blockers with this though, so I'll defer to Tom for official sign off 👍

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt:44]: We need to add docs for all public functions, otherwise detekt fails. Think it doesn't harm, it actually helps to understand that we are fetching info for the current tab and returns null in case of error.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStore.kt
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckStore.kt:44]: I think we might do that soon as detekt complains about it, I'll hold off till then as we'll get a better picture then 🖼️

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt:48]: 😅 yeah was bothering me too but it was the IDE formatter and ktlint in combo that shaped this. Just updated to import the actions so it's in one line 😄

File: fenix/app/src/test/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckServiceImplTest.kt
Comment (rahulsainani) [reply fenix/app/src/test/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckServiceImplTest.kt:28]: So seems like you're referencing the 3rd test in this file, basically it's testing that we are infact using the selected tab url, if not the request would fail or worse, show a different product analysis. So now if I change the implementation to `firstTab` or something else, this test will catch it.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/di/ReviewQualityCheckMiddlewareProvider.kt
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/di/ReviewQualityCheckMiddlewareProvider.kt]: thanks, updated to add more clarity 👍

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckAction.kt
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckAction.kt:25]: Hmm yeah, removed `MiddlewareAction` now that we're not using it. and the other 3 are grouped together

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt
Comment (t-p-white) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt:41]: My thoughts were that this is identical to the interface doc which this fun will use if not doc'd & I thought detekt recognized when a fun was an 'override' so it wouldn't fail. Totally fine to keep it, just wanted to give context to my thoughts 🙂

**Review by t-p-white — APPROVED** [review]
Nice work! Nothing blocking from me

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt:41]: Oh I see, good point! thanks for the explanation, TIL!

**Pull request merged**
